### DISCUSSION TITLE
New at Python - downloading a list of friends from a list of accounts

### DISCUSSION THREAD

**Discussion by DV777** [tweepy/tweepy#1600] [Questions] [answered]
Hello, as mentioned in the title I am entirely new at Python, basically know how to adapt scripts, create loops and using pandas... 

I am trying to obtain the complete list of friends and followers from a specific set of accounts as part of my research project (I am a PhD student). Is it possible with Tweepy ? I really don't get how to do it from the documentation and feel quite dumb each time I look at it.

I have my tokens (I generated a standalone app) but apart from authenticating myself, I cannot do anything with Tweepy...

Thanks in advance to the (very) patient person who will manage to help me.
```
access_token='*'
access_token_secret='*'
consumer_key='*'
consumer_secret='*'
bearer_token='*'
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
api = tweepy.API(auth, wait_on_rate_limit=True)
try:
    redirect_url = auth.get_authorization_url()
except tweepy.TweepError:
    print('Error! Failed to get request token.')
api.update_status('tweepy + oauth!')

def limit_handled(cursor):
    while True:
        try:
            yield cursor.next()
        except tweepy.RateLimitError:
            time.sleep(15 * 60)

for friend in tweepy.Cursor(api.friends).items():
    # Process the friend here
    process_friend(friend)   

for follower in limit_handled(tweepy.Cursor(api.followers).items()):
    print(follower.screen_name)

**Comment by Harmon758** [comment]
> Is it possible with Tweepy ?

Yes

Is there a specific issue you're encountering? What is `process_friend`?

**Comment by DV777** [comment]
> > Is it possible with Tweepy ?
> 
> Yes
> 
> Is there a specific issue you're encountering? What is `process_friend`?

Actually I cannot manage to write a script to scrap all these friends and followers. "process_friend" was in the original example provided in the doc, I thought it was mandatory to include it smh but it's actually just an undefined variable 👍

**Comment by Harmon758** [comment, marked as answer]
From this, `process_friend` is an undefined function, yes.
Otherwise, this is fine for iterating through friends and followers.
Feel free to ask about any specific issues you have.

**Reply by DV777** [reply]
Let me warn you again, I really understand close to nothing in script edition... 😅 

Well if this is supposed to work to collect friends, first where do I find them ? It gives no data. How do I transfer them to a csv or json file ? 
How do I change the collection from my account to a list of accounts I'd like to collect friends from ? (I know how to write a loop, but I cannot see the parameters to change here...)

**Reply by Harmon758** [reply]
[`API.friends`](https://tweepy.readthedocs.io/en/v3.10.0/api.html#API.friends) returns a list of [User objects](https://developer.twitter.com/en/docs/twitter-api/v1/data-dictionary/object-model/user).
You're iterating through each object/model with `for friend in tweepy.Cursor(api.friends).items():`.
What you do with them and how you use each one is up to you.

`API.friends` accepts a user ID or screen name through its `user_id` and `screen_name` parameters, respectively.

**Reply by DV777** [reply]
Ok, where should I add these parameters ? in items ?
How do I interact with them afterwards ? 
`for friend in tweepy.Cursor(api.friends).items()`

**Reply by Harmon758** [reply]
> Ok, where should I add these parameters ? in items ?

Where you initialize the cursor.
You should see the [Cursor Tutorial](https://tweepy.readthedocs.io/en/v3.10.0/cursor_tutorial.html) in the documentation for v3.10.0 and/or the [Pagination](https://tweepy.readthedocs.io/en/latest/pagination.html) section of the latest documentation.

> How do I interact with them afterwards ?

I'm not sure what you're asking. Interact with what?
If you mean the User objects, `friend` is each one you're iterating through in the loop.

**Reply by DV777** [reply]
For example, the code that I forwarded is not working on my side - I keep getting errors (last one : `TweepError: [{'code': 187, 'message': 'Status is a duplicate.'}]`)

How to make it work ?
How to obtain a json or csv file with the list of friends and followers collected ?

**Reply by Harmon758** [reply]
If you look at the full traceback for that error, it's likely pointing to `api.update_status('tweepy + oauth!')`, where you're probably attempting to Tweet something exactly the same as a Tweet you recently Tweeted.

> How to obtain a json or csv file with the list of friends and followers collected ?

You'll need to populate either yourself as you iterate through the User objects.

**Reply by DV777** [reply]
> If you look at the full traceback for that error, it's likely pointing to `api.update_status('tweepy + oauth!')`, where you're probably attempting to Tweet something exactly the same as a Tweet you recently Tweeted.
> 
> > How to obtain a json or csv file with the list of friends and followers collected ?
> 
> You'll need to populate either yourself as you iterate through the User objects.

Thanks ! Yes you were right. 😓 

What do you call populating either through the user objects ? 

What would this error mean ?

`StopIteration
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "<ipython-input-19-516747a0d155>", line 4, in <module>
    for follower in limit_handled(tweepy.Cursor(api.followers).items()):
RuntimeError: generator raised StopIteration`

**Reply by Harmon758** [reply]
> What do you call populating either through the user objects ?

Again, you're iterating through the User objects with `for friend in tweepy.Cursor(api.friends).items()`. You can use each User object however you want, including by adding whatever data from the User object you want into a CSV or JSON file.

> What would this error mean ?

Your `limit_handled` generator doesn't properly handle the `StopIteration` raised by the iterator when all the followers have been iterated through. See [PEP 479](https://www.python.org/dev/peps/pep-0479/).

For code block usage, see https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks.

**Reply by DV777** [reply]
> > What do you call populating either through the user objects ?
> 
> Again, you're iterating through the User objects with `for friend in tweepy.Cursor(api.friends).items()`. You can use each User object however you want, including by adding whatever data from the User object you want into a CSV or JSON file.
> 
> > What would this error mean ?
> 
> Your `limit_handled` generator doesn't properly handle the `StopIteration` raised by the iterator when all the followers have been iterated through. See [PEP 479](https://www.python.org/dev/peps/pep-0479/).
> 
> For code block usage, see https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks.

I'll try to look more into it but all this is way more advanced than my current level. The links you sent are talking a language I can read but not understand. I will try to fix my code today again, can I try to reformulate my questions here in case I am not successful ? (99% chance of that happening tbh..)

**Reply by Harmon758** [reply]
Yes, feel free to ask about any specific issues you encounter.
